src/WebJobs.Script/Binding/Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -201,14 +201,7 @@ internal static void ApplyBindingData(object target, IDictionary<string, object>
                         propertyHelper.Property.CanWrite)
                     {
                         object value = pair.Value;
-                        Type targetType = propertyHelper.Property.PropertyType;
-                        if (value != null && value.GetType() != targetType)
```

Comment:
This was the bug - we were doing type equality, rather than IsAssignableFrom. The property is of type IDictionary but the deserialized value was Dictionary, causing the inequality to pass, and the ChangeType to fail.